### PR TITLE
DDP examples (Single and Multi-node)

### PR DESCRIPTION
**Pull request by Blaizzy**
# Description

This example demonstrates how to use Neptune when running single or multi-node DDP training

---

## ❔ This change

- [x] adds a new feature
- [ ] fixes breaking code
- [ ] is cosmetic (refactoring/reformatting)

---

## ✔️ Pre-merge checklist

Mark which items have been completed:

- [x] Refactored code ([sourcery](https://sourcery.ai/))
- [x] Tested code locally
- [x] Added code to GitHub tests ([notebooks](workflows/test-notebooks.yml), [scripts](workflows/test-scripts.yml))
- [x] Updated GitHub [README](../README.md)

### 🪝 Run the following to both [pre-commit](https://pre-commit.com/) hooks and JupyterLab (turn on `format on save`)

- [x] Stripped outputs from notebooks - ([nbstripout](https://pypi.org/project/nbstripout/))
- [x] Sorted imports - ([isort](https://pycqa.github.io/isort/))
- [x] Formatted code - ([back](https://github.com/psf/black))

### 📄 Update documentation

Coordinate with @normandy7

- [ ] Updated content in the [docs](https://docs.neptune.ai)
- [ ] Added to all relevant docs index pages (e.g., [Examples](https://docs.neptune.ai/getting-started/examples), [Integrations intro](https://docs.neptune.ai/integrations-and-supported-tools/intro), [Model training index](https://docs.neptune.ai/integrations-and-supported-tools/model-training))
- [ ] Added to examples [README](../README.md)

---

## 🧪 Test Configuration

- OS:
- Python version:
- Neptune version:
- Affected libraries with version:

---

## ✔️ Post-merge checklist

These tasks should be done after the code is merged to main. Create a subtask under the main ClickUp task to have these on your radar and share the subtask here.

__Post-merge subtask:__ `(ClickUp/JIRA task name)`

- [x] Update relevant blogs with impacted code/functionality (Contact @Patrycja-J for the list of blogs)
- [ ] Add functionality under the Neptune column of the `competitor comparison sheet` (Coordinate with @SiddhantSadangi)
- [ ] Update integration README (Contact @Patrycja-J)

### CODE REVIEW

**Comment by Blaizzy** [comment]
@SiddhantSadangi 

Remarks:
- No notebooks, as this is not a must in this use case. Data shows that most users run scripts, plus Colab doesn't really give you multi-GPU for free.
- No shell script since we can't run tests with GPU in github actions.

**Comment by Blaizzy** [comment]
For now, no action is needed. 

I'm going to run some final tests in the morning and give you the green light after that.

**Comment by Blaizzy** [comment]
@SiddhantSadangi ready ✅

**Comment by Blaizzy** [comment]
@normandy7 here is are the DDP code examples

**Comment by SiddhantSadangi** [comment]
> No shell script since we can't run tests with GPU in github actions.

Adding GPU runners is in the GitHub roadmap (https://github.com/github/roadmap/issues/505).. lets have the scripts so that we can start testing as soon as this is added to GH

**Comment by Blaizzy** [comment]
> Adding GPU runners is in the GitHub roadmap (https://github.com/github/roadmap/issues/505).. let's have the scripts so that we can start testing as soon as this is added to GH

Sure, I can do that. But there won't be one for the multi-node example.

**Comment by Blaizzy** [comment]
Added a shell example for single-node and 2 GPUs

File: how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        shuffle=False,
+        sampler=test_sampler,
+        num_workers=4,
+    )
+    return trainloader, testloader
+
+
+def train(net, trainloader, run, rank, params):
+    if rank == 0:
+        # Log params
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py:51]:
```suggestion
        # (neptune) log params
```

File: how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            # Gather loss value from all processes on main process for logging
+            dist.reduce(tensor=loss, dst=0)
+            dist.barrier()  # synchronizes all the threads
+
+            if rank == 0:
+                running_loss += loss.item() / dist.get_world_size()
+
+        if rank == 0:
+            epoch_loss = running_loss / num_of_batches
+            # Log metrics
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py:79]:
```suggestion
            # (neptune) log metrics
```

File: how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+            dist.reduce(tensor=predicted, dst=0)
+            dist.barrier()  # synchronizes all the threads
+
+            if rank == 0:
+                correct += (predicted == labels).sum().item()
+
+    if rank == 0:
+        acc = 100 * correct // total
+        # Log metrics
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py:111]:
```suggestion
        # (neptune) log metrics
```

File: how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    net = torchvision.models.resnet50(weights=None).to(f"cuda:{rank}")
+    net = nn.SyncBatchNorm.convert_sync_batchnorm(net)
+    net = nn.parallel.DistributedDataParallel(net, device_ids=[rank])
+
+    # To correctly monitor each GPU usage
+    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+    os.environ["CUDA_VISIBLE_DEVICES"] = str(rank)
+
+    # Creates multiple run instances
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multi_node_multiple_runs.py:149]:
```suggestion
    # (neptune) create multiple run instances
```

File: how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        sampler=test_sampler,
+        num_workers=4,
+    )
+    return trainloader, testloader
+
+
+def train(net, trainloader, run, rank, params):
+
+    if rank == 0:
+        # Log params
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py:54]:
```suggestion
        # (neptune) log params
```

File: how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            optimizer.zero_grad()
+            outputs = net(images)
+            loss = criterion(outputs, labels)
+            loss.backward()
+            optimizer.step()
+            running_loss += loss.item()
+
+        if rank == 0:
+            epoch_loss = running_loss / num_of_batches
+            # Log metrics
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py:76]:
```suggestion
            # (neptune) log metrics
```

File: how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            images, labels = data
+            images, labels = images.cuda(), labels.cuda()
+            outputs = net(images)
+            _, predicted = torch.max(outputs.data, 1)
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
+
+    if rank == 0:
+        acc = 100 * correct // total
+        # Log metrics
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py:99]:
```suggestion
        # (neptune) log metrics
```

File: how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+    net = torchvision.models.resnet50(weights=None).cuda()
+    net = nn.SyncBatchNorm.convert_sync_batchnorm(net)
+    net = nn.parallel.DistributedDataParallel(net, device_ids=[local_rank])
+
+    # To correctly monitor each GPU usage
+    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+    os.environ["CUDA_VISIBLE_DEVICES"] = str(rank)
+
+    # Automatically create and broadcast `custom_run_id` to all processes
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py:141]:
```suggestion
    # (neptune) automatically create and broadcast `custom_run_id` to all processes
```

File: how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        custom_run_id = [hashlib.md5(str(time.time()).encode()).hexdigest()]
+        monitoring_namespace = "monitoring"
+    else:
+        custom_run_id = [None]
+        monitoring_namespace = f"monitoring/{rank}"
+
+    dist.broadcast_object_list(custom_run_id, src=0)
+    custom_run_id = custom_run_id[0]
+
+    # Creates multiple run instances
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py:152]:
```suggestion
    # (neptune) create multiple run instances
```

File: how-to-guides/ddp-training/scripts/train_ddp_single_run.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            optimizer.zero_grad()
+            outputs = net(images)
+            loss = criterion(outputs, labels)
+            loss.backward()
+            optimizer.step()
+            running_loss += loss.item()
+
+        if rank == 0:
+            epoch_loss = running_loss / num_of_batches
+            # Log metrics
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_single_run.py:74]:
```suggestion
            # (neptune) log metrics
```

File: how-to-guides/ddp-training/scripts/train_ddp_single_run.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            images, labels = data
+            images, labels = images.cuda(), labels.cuda()
+            outputs = net(images)
+            _, predicted = torch.max(outputs.data, 1)
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
+
+    if rank == 0:
+        acc = 100 * correct // total
+        # Log metrics
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_single_run.py:97]:
```suggestion
        # (neptune) log metrics
```

File: how-to-guides/ddp-training/scripts/train_ddp_single_run.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+    net = torchvision.models.resnet50(weights=None).cuda()
+    net = nn.SyncBatchNorm.convert_sync_batchnorm(net)
+    net = nn.parallel.DistributedDataParallel(net, device_ids=[local_rank])
+
+    # To correctly monitor each GPU usage
+    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+    os.environ["CUDA_VISIBLE_DEVICES"] = "0,1"  # No. of GPUs installed
+
+    # Create new run on the main process only
```
Comment (SiddhantSadangi) [inline how-to-guides/ddp-training/scripts/train_ddp_single_run.py:139]:
```suggestion
    # (neptune) create new run on the main process only
```

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    # But all instances log metadata to the same run
+    # by passing the `custom_run_id` argument
```
[the remark this answers is not in the document]
Comment (SiddhantSadangi) [reply how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py:154]: We have `(neptune)` (all-lowers) in other code pieces as well.. not sure why though.
Can adopt `(Neptune)` starting here.

File: how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py (hunk 8 above)
Comment (Blaizzy) [reply how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py:141]: @normandy7  

`(neptune) log...` vs. `(neptune) Log...` vs. `(Neptune) Log...` vs. `(Neptune) log...`

Comment (normandy7) [reply how-to-guides/ddp-training/scripts/train_ddp_multiple_runs.py]: `(Neptune) Log...` -- assuming "Log" is what starts a complete sentence. I'd always use sentence case for sentence-like comments.

**Comment by Blaizzy** [comment]
@SiddhantSadangi I don't have permission to resolve this conflict but I have done via commits

**Comment by SiddhantSadangi** [comment]
> @SiddhantSadangi I don't have permission to resolve this conflict, but I have done via commits Screenshot 2022-12-02 at 3 17 23 PM

resolved

**Comment by Blaizzy** [comment]
Done the branch is up to date 👍

**Review by SiddhantSadangi — APPROVED** [review]

**Comment by Blaizzy** [comment]
@SiddhantSadangi @normandy7 anything missing here?

**Comment by SiddhantSadangi** [comment]
> @SiddhantSadangi @normandy7 anything missing here?

all good from my end. Waiting for @normandy7's approval

**Review by normandy7 — APPROVED** [review]

**Comment by SiddhantSadangi** [comment]
@Blaizzy - looks like we missed adding ddp code to tests.
Can you update the workflow files?

**Comment by SiddhantSadangi** [comment]
> > No shell script since we can't run tests with GPU in github actions.
> 
> Adding GPU runners is in the GitHub roadmap ([github/roadmap#505](https://github.com/github/roadmap/issues/505)).. lets have the scripts so that we can start testing as soon as this is added to GH

@Blaizzy - lets add the ddp examples to the tests so that we have coverage once GPUs are supported, it might slip through the cracks otherwise

**Pull request merged**